bigquery/v2/query/query.go

```diff
@@ -0,0 +1,183 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package query
+
+import (
+	"context"
+	"time"
+
+	"cloud.google.com/go/bigquery/v2/apiv2/bigquerypb"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+// Query represents a query job.
+type Query struct {
+	c         *Client
+	complete  bool
+	projectID string
+	jobID     string
+	location  string
+
+	cachedTotalRows uint64
+	cachedPageToken string
+	cachedRows      []*Row
+	cachedSchema    *schema
+}
+
+func newQueryJobFromQueryResponse(c *Client, res *bigquerypb.QueryResponse) (*Query, error) {
+	schema := newSchema(res.Schema)
+	q := &Query{
+		c:               c,
+		cachedSchema:    schema,
+		cachedPageToken: res.PageToken,
+	}
+	if res.TotalRows != nil {
+		q.cachedTotalRows = res.TotalRows.Value
+	}
+	if res.JobComplete != nil {
+		q.complete = res.JobComplete.Value
+	}
+	if res.Rows != nil {
+		var err error
+		q.cachedRows, err = fieldValueRowsToRowList(res.Rows, schema)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if res.JobReference != nil {
+		jobRef := res.JobReference
+		q.projectID = jobRef.ProjectId
+		q.jobID = jobRef.JobId
+		if jobRef.Location != nil {
+			q.location = jobRef.Location.GetValue()
+		}
+	}
+	return q, nil
+}
+
+func newQueryJobFromJob(c *Client, job *bigquerypb.Job) (*Query, error) {
+	return newQueryJobFromJobReference(c, nil, job.JobReference)
+}
+
+func newQueryJobFromJobReference(c *Client, schema *bigquerypb.TableSchema, jobRef *bigquerypb.JobReference) (*Query, error) {
+	res := &bigquerypb.QueryResponse{
+		Schema:       schema,
+		JobReference: jobRef,
+	}
+	return newQueryJobFromQueryResponse(c, res)
+}
+
+// state is one of a sequence of states that a Job progresses through as it is processed.
+type state = string
+
+const (
+	// Pending is a state that describes that the job is pending.
+	Pending state = "PENDING"
+	// Running is a state that describes that the job is running.
+	Running state = "RUNNING"
+	// Done is a state that describes that the job is done.
+	Done state = "DONE"
+)
+
+func (q *Query) checkStatus(ctx context.Context) error {
```

Comment:
this fn name confuses me a bit.  The implementation is a bit wonky since it explicitly asks for no rows but appears to have some row cache logic embedded via consumeQueryResponse.